zk_prover/src/chips/less_than_vertical.rs

```diff
@@ -0,0 +1,204 @@
+/* This is a 'vertical' implementation of LTChip
+It reduces the number of advice columns present in the original 'horizontal approach */
+
+use halo2_proofs::{
+    circuit::{Chip, Layouter, Region, Value},
+    halo2curves::{bn256::Fr as Fp, ff::PrimeField},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Fixed, VirtualCells},
+    poly::Rotation,
+};
+
+use gadgets::{
+    bool_check,
+    util::{expr_from_bytes, pow_of_two},
+};
+
+/// Instruction that the Lt vertical chip needs to implement.
+pub trait LtVerticalInstruction {
+    /// Assign the lhs and rhs witnesses to the Lt chip's region.
+    fn assign(
+        &self,
+        region: &mut Region<'_, Fp>,
+        offset: usize,
+        lhs: Value<Fp>,
+        rhs: Value<Fp>,
+    ) -> Result<(), Error>;
+
+    /// Load the u8 lookup table.
+    fn load(&self, layouter: &mut impl Layouter<Fp>) -> Result<(), Error>;
+}
+
+/// Config for the LtVertical chip.
+#[derive(Clone, Copy, Debug)]
+pub struct LtVerticalConfig<const N_BYTES: usize> {
+    /// Denotes the lt outcome. If lhs < rhs then lt == 1, otherwise lt == 0.
+    pub lt: Column<Advice>,
+    /// Denotes the bytes representation of the difference between lhs and rhs.
+    pub diff: Column<Advice>,
+    /// Denotes the range within which each byte should lie.
+    pub u8: Column<Fixed>,
+    /// Denotes the range within which both lhs and rhs lie.
+    pub range: Fp,
+}
+
+impl<const N_BYTES: usize> LtVerticalConfig<N_BYTES> {
+    /// Returns an expression that denotes whether lhs < rhs, or not.
+    pub fn is_lt(&self, meta: &mut VirtualCells<Fp>, rotation: Option<Rotation>) -> Expression<Fp> {
+        meta.query_advice(self.lt, rotation.unwrap_or_else(Rotation::cur))
+    }
+}
+
+/// Chip that compares lhs < rhs.
+#[derive(Clone, Debug)]
+pub struct LtVerticalChip<const N_BYTES: usize> {
+    config: LtVerticalConfig<N_BYTES>,
+}
+
+impl<const N_BYTES: usize> LtVerticalChip<N_BYTES> {
+    /// Configures the LtVertical chip.
+    pub fn configure(
+        meta: &mut ConstraintSystem<Fp>,
+        q_enable: impl FnOnce(&mut VirtualCells<'_, Fp>) -> Expression<Fp>,
+        lhs: impl FnOnce(&mut VirtualCells<Fp>) -> Expression<Fp>,
+        rhs: impl FnOnce(&mut VirtualCells<Fp>) -> Expression<Fp>,
+    ) -> LtVerticalConfig<N_BYTES> {
+        let lt = meta.advice_column();
+        let diff = meta.advice_column();
+        let range = pow_of_two(N_BYTES * 8);
+        let u8 = meta.fixed_column();
+
+        meta.create_gate("lt gate", |meta| {
+            let q_enable = q_enable(meta);
+            let lt = meta.query_advice(lt, Rotation::cur());
+
+            let mut diff_bytes = Vec::<Expression<Fp>>::new();
+
+            diff_bytes.push(meta.query_advice(diff, Rotation::cur()));
+            diff_bytes.push(meta.query_advice(diff, Rotation::next()));
+
+            for i in 2..N_BYTES {
+                diff_bytes.push(meta.query_advice(diff, Rotation(i as i32)));
```

Comment:
I believe that this can be as a for loop from 0 to N_Bytes. Rotation(0) should be equivalent to `Rotation::cur()`.